### PR TITLE
qt5.qtbase: ignore hook propagated from buildPlatform

### PR DESCRIPTION
**Pull request by ghost**
Includes

- https://github.com/NixOS/nixpkgs/pull/220518

## Description of changes

The main headache with cross-compiling QT is that our QT derivations go bananas with `propagatedInputs`.  Unfortunately this means that both `buildPackages.qtbase` and `pkgs.qtbase` get propagated into the dependencies of every cross-compiled build, which triggers the hook check:

https://github.com/NixOS/nixpkgs/blob/feaa219f3988126d4241055df129f322ae63dcdf/pkgs/development/libraries/qt-5/hooks/qtbase-setup-hook.sh#L1-L9

Previously in https://github.com/NixOS/nixpkgs/pull/227900 I simply disabled the check outright for cross-compiled builds.  This is a more refined solution, where we skip the hook if has a different `stdenv.outPath` than the derivation being built.

This works because the outpath of `stdenv` is written into the environment of *every* derivation.  There's nothing else we can use that's guaranteed to be in *every* derivation that `qtbase-setup-hook.sh` is propagated into.

We should stop using `propagatedBuildInputs`.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/qt-5/hooks/qtbase-setup-hook.sh [hunk 1]
```diff
@@ -1,3 +1,9 @@
+if (( "$targetOffset" <= 0 )); then
```
Comment (Artturin) [inline pkgs/development/libraries/qt-5/hooks/qtbase-setup-hook.sh:1]: Also follow strictDeps like cc-wrapper? https://github.com/NixOS/nixpkgs/blob/d7abb3a970302200f7a57075d64f837d855bdeba/pkgs/build-support/cc-wrapper/setup-hook.sh#L61

Comment (amjoseph-nixpkgs) [reply pkgs/development/libraries/qt-5/hooks/qtbase-setup-hook.sh:1]: No, I think it was a mistake to have `strictDeps` be opt-in.

Also, the `qtbase-setup-hook` is only allowed to run once per dependency outpath, but a dependency outpath can appear more than once with *different $hostOffset values*.

I think this check is too important to allow it to be muted by default.

Comment (amjoseph-nixpkgs) [reply pkgs/development/libraries/qt-5/hooks/qtbase-setup-hook.sh:1]: > Also, the `qtbase-setup-hook` is only allowed to run once per dependency outpath, but a dependency outpath can appear more than once with _different $hostOffset values_.

This means that the existing code is very sensitive to the order that dependencies appear in.  It's really bad.

Comment (Artturin) [reply pkgs/development/libraries/qt-5/hooks/qtbase-setup-hook.sh:1]: My bad, I thought the strictDeps part would make it stricter in some way, bash conditional....

Comment (amjoseph-nixpkgs) [reply pkgs/development/libraries/qt-5/hooks/qtbase-setup-hook.sh]: I found a better solution: compare the `stdenv` environment variable.  See latest push  e722e3e

**Comment by amjoseph-nixpkgs** [comment]
```
error: building of '/nix/store/73fp0qc16cfbn2ddhfxg9b564zix7242-llvm-16.0.6.drv!dev' from .drv file timed out after 3600 seconds
```

**Comment by amjoseph-nixpkgs** [comment]
Rebased, maybe this time LLVM won't time out.

File: pkgs/development/libraries/qt-5/5.15/default.nix [hunk 2]
```diff
@@ -353,7 +351,8 @@ let
 
       wrapQtAppsHook = callPackage ({ makeBinaryWrapper, qtbase, qtwayland }: makeSetupHook {
         name = "wrap-qt5-apps-hook";
-        propagatedBuildInputs = [ qtbase.dev makeBinaryWrapper ]
+        propagatedBuildInputs = [ makeBinaryWrapper ];
+        depsTargetTargetPropagated = [ qtbase.dev ]
```
Comment (Artturin) [inline pkgs/development/libraries/qt-5/5.15/default.nix:355]: what about the build tools?

**Comment by Artturin** [comment]
Applying to master

````
qtbase-aarch64-unknown-linux-gnu> qxcbcursor.cpp: In constructor 'QXcbCursor::QXcbCursor(QXcbConnection*, QXcbScreen*)':
qtbase-aarch64-unknown-linux-gnu> qxcbcursor.cpp:320:43: error: 'NIXPKGS_LIBXCURSOR' was not declared in this scope
qtbase-aarch64-unknown-linux-gnu>   320 |         QLibrary xcursorLib(QLatin1String(NIXPKGS_LIBXCURSOR), 1);
qtbase-aarch64-unknown-linux-gnu>       |                                           ^~~~~~~~~~~~~~~~~~
```

Native breaks too?

 https://github.com/NixOS/nixpkgs/blob/67273d0def72ec49e6d4c7225e032a87bbdaaca7/pkgs/development/libraries/qt-5/modules/qtbase.nix#L218

File: pkgs/development/libraries/qt-5/modules/qtbase.nix [hunk 3]
```diff
@@ -391,6 +391,7 @@ stdenv.mkDerivation (finalAttrs: ({
   dontStrip = debugSymbols;
 
   setupHook = ../hooks/qtbase-setup-hook.sh;
+  env.qtbase_stdenv = stdenv.outPath;
```
Comment (Artturin) [inline pkgs/development/libraries/qt-5/modules/qtbase.nix:394]: Ah you override the env

**Pull request closed without merging**